ubuntu/devel: Clean up postinst and preinst scripts

holmanb opened the pull request:
Drop version gated code from preinst.
Add version gates to old code in preinst and postinst.

<!--
Thank you for submitting a PR to cloud-init!

To ease the process of reviewing your PR, do make sure to complete the following checklist **before** submitting a pull request.

- [ ] I have signed the CLA: https://ubuntu.com/legal/contributors
- [ ] I have added my Github username to ``tools/.github-cla-signers``
- [ ] I have included a comprehensive commit message using the guide below
- [ ] I have added unit tests to cover the new behavior under ``tests/unittests/``
  - Test files should map to source files i.e. a source file ``cloudinit/example.py`` should be tested by ``tests/unittests/test_example.py``
  - Run unit tests with ``tox -e py3``
- [ ] I have kept the change small, avoiding unnecessary whitespace or non-functional changes.
- [ ] I have added a reference to issues that this PR relates to in the PR message (Refs GH-1234, Fixes GH-1234)
- [ ] I have updated the documentation with the changed behavior.
  - If the change doesn't change the user interface and is trivial, this step may be skipped.
  - Cloud-config documentation is generated from the jsonschema.
  - Generate docs with ``tox -e docs``.
-->


## Proposed Commit Message
<!-- Include a proposed commit message because PRs are squash merged
by default.

See https://www.conventionalcommits.org/en/v1.0.0/#specification
for our commit message convention.

If the change is related to a particular cloud or particular distro,
please include the "optional scope" in the summary line. E.g.,
feat(ec2): Add support for foo to the baz

Types used by this project:
feat, fix, docs, ci, test, refactor, chore
-->
```
chore(debian): Clean up postinst and preinst scripts

Drop version gated code from preinst.
Add version gates to old code in preinst and postinst.
```

## Additional Context
<!-- If relevant -->

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->


## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

holmanb commented on debian/cloud-init.postinst:
```diff
@@ -7,54 +7,12 @@ set -e
 fix_1336855() {
   ### Begin fix for LP: 1336855
-  # fix issue where cloud-init misidentifies the location of grub and
+  # fix issue where cloud-init misidentified the location of grub and
```
This isn't relevant anymore for images initialized using cloud_style_initialization, and I'm not convinced that we want it on those that aren't, either.

holmanb replied:
Hmm, true. My approach was unnecessarily cautious.

TheRealFalcon commented on debian/cloud-init.preinst:
```diff
@@ -1,237 +1,39 @@
    # Remove vestigial /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg
-   # from Oracle now that datasource honors system cfg above datasource cfg.
+   local oldver="$1" last_bad_ver="24.3~1g6e4153b3-0ubuntu1"
+   dpkg --compare-versions "$oldver" le "$last_bad_ver" || return 0  # from Oracle now that datasource honors system cfg above datasource cfg.
```
This comment got moved incorrectly.

TheRealFalcon commented on debian/cloud-init.preinst:
```diff
@@ -1,237 +1,39 @@
 
+cleanup_user_scripts() {
+   local oldver="$1" last_bad_ver="24.3~1g6e4153b3-0ubuntu1"
+   dpkg --compare-versions "$oldver" le "$last_bad_ver" || return 0  # from Oracle now that datasource honors system cfg above datasource cfg.
```
This comment shouldn't be here.

TheRealFalcon commented on debian/cloud-init.preinst:
```diff
@@ -1,237 +1,39 @@
    rm -f /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg
 }
 
+cleanup_user_scripts() {
```
Why is the contents of this function needed compared to the other pieces that were removed. The referenced LP is from 2012.

holmanb replied:
Agreed. I was thinking about this in the wrong way.

TheRealFalcon commented on debian/cloud-init.preinst:
```diff
@@ -1,237 +1,39 @@
+   fi
+}
+
+cleanup_rename_config() {
```
Any reason we can't also kill this? I think even back to Trusty we have `90_dpkg.cfg` and no `distro.cfg`.

holmanb replied:
Agreed

blackboxsw reviewed (changes requested):
We need to document these changes in a new  UNRELEASED d/changelog entry.
Given that we had released to oracular, we'll probably need to rebase on latest upstream/ubuntu/devel

shellcheck warnings on the scripts don't generate any new warning conditions for non-POSIX patterns we don't already use in other functions.

Upgrade path on oracular doesn't generate any error conditions.

Unrelated to this PR, lintian is not thrilled with a now bogus/deprecated/retired override
```
 lintian  cloud-init_24.4~2g2e4c39b7-0ubuntu2_all.deb
E: cloud-init: alien-tag package-supports-alternative-init-but-no-init.d-script [usr/share/lintian/overrides/cloud-init:7]
lintian -i cloud-init_24.4~2g2e4c39b7-0ubuntu2_all.deb
N:
E: cloud-init: alien-tag package-supports-alternative-init-but-no-init.d-script [usr/share/lintian/overrides/cloud-init:7]
N: 
N:   The given override refers to an unknown tag.
N: 
N:   Please refer to Format of override files (Section 2.4.1) in the Lintian
N:   User's Manual for details.
N: 
N:   Visibility: error
N:   Show-Always: yes
N:   Check: debian/lintian-overrides/mystery
```

We can delete  that speciifc lintian override in this PR if you want (and we may need to talk separately about lintian warnings
```
W: cloud-init: systemd-service-file-refers-to-unusual-wantedby-target cloud-init.target [lib/systemd/system/cloud-init-network.service]
N:
W: cloud-init: systemd-service-file-shutdown-problems [lib/systemd/system/cloud-init-main.service]
N: 
N:   The specified systemd .service file contains both DefaultDependencies=no
N:   and Conflicts=shutdown.target directives without Before=shutdown.target.
N:   
N:   This can lead to problems during shutdown because the service may linger
N:   until the very end of shutdown sequence as nothing requests to stop it
N:   before (due to DefaultDependencies=no).
N:   
N:   There is race condition between stopping units and systemd getting a
N:   request to exit the main loop, so it may proceed with shutdown before all
N:   pending stop jobs have been processed.
N:   
N:   Please add Before=shutdown.target.
N: 
N:   Please refer to https://github.com/systemd/systemd/issues/11821 for
N:   details.
N: 
N:   Visibility: warning
N:   Show-Always: no
N:   Check: systemd
N: 
N: 0 hints overridden; 1 unused override
```


```

holmanb commented on debian/cloud-init.postinst:
```diff
@@ -7,81 +7,4 @@ set -e
 db_capb escape # to support carriage return / multi-line values
 
-fix_lp1889555() {
-    local oldver="$1" last_bad_ver="20.3-2-g371b392c-0ubuntu1"
```
Jammy's first release was `22.1-14-g2e17a0d6-0ubuntu1~22.04.5`. This should be safe to drop.

holmanb commented on debian/cloud-init.postinst:
```diff
@@ -7,81 +7,4 @@ set -e
-}
-
-update_maas_spelling(){
-   local oldver="$1" last_bad_ver="24.3~1g6e4153b3-0ubuntu1"
```
This maas spelling has existed since xenial (maybe before), so this shouldn't be needed anymore.

holmanb commented on debian/cloud-init.postrm:
```diff
@@ -5,10 +5,6 @@ set -e
 case "$1" in
    purge)
       rm -f /etc/cloud/cloud.cfg.d/90_dpkg.cfg
-      rm -f /etc/apt/apt.conf.d/90cloud-init-pipelining
```
Cloud-init's used to have a matching postinst which installed this file (it ran `cloud-init single` from the postinst). Cloud-init no longer automatically installs it, so we can drop this.

blackboxsw replied:
ack. Dropped in cloud-init 18.5, which means any  package upgrades coming from bionic++ will already have this patch.

holmanb commented on debian/cloud-init.prerm:
```diff
@@ -1,6 +0,0 @@
-#!/bin/sh
-
-set -e
-rm -f /etc/cron.d/cloudinit-updates
```
This hasn't been used in a very long time.

blackboxsw replied:
Agreed, pre version 0.7.7

holmanb commented:
> ```
> W: cloud-init: systemd-service-file-refers-to-unusual-wantedby-target cloud-init.target [lib/systemd/system/cloud-init-network.service]
> N:
> W: cloud-init: systemd-service-file-shutdown-problems [lib/systemd/system/cloud-init-main.service]
> N: 
> N:   The specified systemd .service file contains both DefaultDependencies=no
> N:   and Conflicts=shutdown.target directives without Before=shutdown.target.
> N:   
> N:   This can lead to problems during shutdown because the service may linger
> N:   until the very end of shutdown sequence as nothing requests to stop it
> N:   before (due to DefaultDependencies=no).
> N:   
> N:   There is race condition between stopping units and systemd getting a
> N:   request to exit the main loop, so it may proceed with shutdown before all
> N:   pending stop jobs have been processed.
> N:   
> N:   Please add Before=shutdown.target.
> N: 
> N:   Please refer to https://github.com/systemd/systemd/issues/11821 for
> N:   details.
> N: 
> N:   Visibility: warning
> N:   Show-Always: no
> N:   Check: systemd
> N: 
> N: 0 hints overridden; 1 unused override
> ```

@blackboxsw I addressed this in https://github.com/canonical/cloud-init/pull/5659

holmanb commented:
@blackboxsw @TheRealFalcon Thanks for the reviews. I just updated the PR to address your comments. I also spotted a couple of more things that I think we can drop which I've added in a separate commit.

aciba90 commented:
Sorry for the conflicts, some lintian issues were addressed in https://github.com/canonical/cloud-init/pull/5652 and I did not see the overlap here.

holmanb commented:
> Sorry for the conflicts, some lintian issues were addressed in #5652 and I did not see the overlap here.

all good, no worries

TheRealFalcon commented:
+1 to changes, but still needs a d/changelog entry.

blackboxsw reviewed (approved):
Approve pending d/changelog entry

holmanb commented:
I just squashed the extra commits. It looks like your "request changes" review hasn't been dismissed by Github yet @TheRealFalcon so I can't merge yet.

TheRealFalcon reviewed (approved):
LGTM!

blackboxsw commented:
@holmanb cleared to land this in ubuntu/devel. oracular accepted our 24.4~3+really24.3.1 so we can rebase and land this unreleased content above that.

aciba90 commented:
Worth linking to https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/2080892 somewhere?
